packages/db/src/query2/builder/functions.ts

```diff
@@ -0,0 +1,392 @@
+import { Agg, Func } from "../ir"
+import { toExpression } from "./ref-proxy.js"
+import type { Expression } from "../ir"
+import type { RefProxy } from "./ref-proxy.js"
+
+// Helper types for type-safe expressions - cleaned up
+
+// Helper type for string operations
+type StringLike<T> =
+  T extends RefProxy<string>
+    ? RefProxy<string> | string | Expression<string>
+    : T extends string
+      ? string | Expression<string>
+      : Expression<string>
+
+// Helper type for numeric operations
+type NumberLike<T> =
+  T extends RefProxy<number>
+    ? RefProxy<number> | number | Expression<number>
+    : T extends number
+      ? number | Expression<number>
+      : Expression<number>
+
+// Helper type for any expression-like value
+type ExpressionLike = Expression | RefProxy<any> | any
+
+// Operators
+
+export function eq(
+  left: RefProxy<string>,
+  right: string | RefProxy<string> | Expression<string>
+): Expression<boolean>
+export function eq(
+  left: RefProxy<number>,
+  right: number | RefProxy<number> | Expression<number>
+): Expression<boolean>
+export function eq(
+  left: RefProxy<boolean>,
+  right: boolean | RefProxy<boolean> | Expression<boolean>
+): Expression<boolean>
+export function eq<T>(
+  left: RefProxy<T>,
+  right: T | RefProxy<T> | Expression<T>
+): Expression<boolean>
+export function eq(
+  left: string,
+  right: string | Expression<string>
+): Expression<boolean>
+export function eq(
+  left: number,
+  right: number | Expression<number>
+): Expression<boolean>
+export function eq(
+  left: boolean,
+  right: boolean | Expression<boolean>
+): Expression<boolean>
+export function eq(
+  left: Expression<string>,
+  right: string | Expression<string>
+): Expression<boolean>
+export function eq(
+  left: Expression<number>,
+  right: number | Expression<number>
+): Expression<boolean>
+export function eq(
+  left: Expression<boolean>,
+  right: boolean | Expression<boolean>
+): Expression<boolean>
+export function eq(left: any, right: any): Expression<boolean> {
+  return new Func(`eq`, [toExpression(left), toExpression(right)])
+}
+
+export function gt(
```

Comment:
These signatures look very similar to `eq`. They can be simplified similarly to how we simplified `eq`. However, conceptually, they are all binary operators and that's why they are so similar. So ideally we could define a generic type signature of binary operators and each binary operator is an instance of that type. That's something we can do if we use `const eq: BinaryOperator<..., ...> = (...) => { ... }` instead of function overloading like here.